subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java

```diff
@@ -72,42 +88,85 @@ public String toString() {
         }
     }
 
-    private static void assertAttributeConstraints(Object value, Attribute<?> attribute) {
+    /**
+     * Checks that the attribute's type matches the given value's type (if a non-{@link Provider}), or the {@link ProviderInternal#getType()} if the
+     * value is a {@link ProviderInternal}; as well as the expectedValueType.
+     *
+     * If the value is a {@link Provider} but <strong>NOT</strong> also
+     * a {@link ProviderInternal}, then no assertions are made (this type of lazily added attribute will fail at the time of retrieval from this container).
+     *
+     * @param value the value (or value provider) to check
+     * @param attribute the attribute containing a type to check against
+     */
+    private void assertAttributeTypeIsValid(@Nullable Object value, Attribute<?> attribute) {
         if (value == null) {
             throw new IllegalArgumentException("Setting null as an attribute value is not allowed");
         }
-        if (!attribute.getType().isAssignableFrom(value.getClass())) {
-            throw new IllegalArgumentException("Unexpected type for attribute '" + attribute.getName() + "'. Expected " + attribute.getType().getName() + " but was:" + value.getClass().getName());
+
+        final Class<?> actualValueType;
+        if (ProviderInternal.class.isAssignableFrom(value.getClass())) {
+            ProviderInternal<?> provider = Cast.uncheckedCast(value);
+            actualValueType = provider.getType();
+        } else if (!Provider.class.isAssignableFrom(value.getClass())) {
+            actualValueType = value.getClass();
+        } else {
+            actualValueType = null;
+        }
+
+        if (null != actualValueType) {
+            if (!attribute.getType().isAssignableFrom(actualValueType)) {
+                throw new IllegalArgumentException("Unexpected type for attribute: '" + attribute.getName() + "'. Attribute type: " + attribute.getType().getName() + " did not match actual type: " + actualValueType.getName());
```

Comment:
I think this doesn't quite read the right way. We say "Unexpected type" and refer to the "actual type" but the message makes it sound like the type we found for the value is the correct type.

I would leave this in terms of expected type, e.g.:

> Unexpected type for attribute 'org.gradle.usage'. Attribute type 'Integer' did not match expected type 'String'.

Where Integer is the type of the value passed in and String is the type the attribute expects.

Reply:
I see why this would be confusing, I'll clarify the message.